### PR TITLE
feat(rf1): Store index ref in metastore

### PR DESCRIPTION
**Pull request by cyriltovena**
**What this PR does / why we need it**:

This stores in the metastore the offset and size of the index for each WAL segment.

We'll be able to use this to avoid downloading all chunks that we don't need in the future read path.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/wal/ref.go [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package wal
+
+type DataRef uint64
+
+func NewDataRef(offset, size uint64) DataRef {
+	return DataRef(offset<<32 | size)
+}
+
+func (b DataRef) Unpack() (int, int) {
```
Comment (benclive) [inline pkg/storage/wal/ref.go:9]: Whats the reason to pack these values?

Comment (cyriltovena) [reply pkg/storage/wal/ref.go:9]: None really, we could use 2 int64 or 2 int32. The code was just available and I copy pasted.

Do you prefer to change it ?

Comment (benclive) [reply pkg/storage/wal/ref.go:9]: If we're using it elsewhere then I'm happy to continue using it. If this is the first usage I'm less inclined to introduce this as a way of storing multiple values unless there is a good reason (maybe due to perf, storage concerns etc.)

File: pkg/storage/wal/ref.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package wal
```
Comment (benclive) [inline pkg/storage/wal/ref.go:1]: this file can be removed now

**Review by benclive — APPROVED** [review]

**Pull request merged**